frontend/src/scenes/billing/BillingLineGraph.tsx

```diff
@@ -0,0 +1,266 @@
+import 'chartjs-adapter-dayjs-3'
+
+import { Chart, ChartDataset, ChartOptions, TooltipModel } from 'chart.js'
+import { getSeriesColor } from 'lib/colors'
+import { useCallback, useEffect, useRef } from 'react'
+import { createRoot, Root } from 'react-dom/client'
+
+import { BillingLineGraphTooltip } from './BillingLineGraphTooltip'
+
+export interface BillingSeriesType {
+    id: number
+    label: string
+    data: number[]
+    dates: string[]
+    valueFormatter?: (value: number) => string
+    showLegend?: boolean
+}
+
+export interface BillingLineGraphProps {
+    series: BillingSeriesType[]
+    dates: string[]
+    isLoading?: boolean
+    hiddenSeries: number[]
+    valueFormatter?: (value: number) => string
+    showLegend?: boolean
+    interval?: 'day' | 'week' | 'month'
+}
+
+const defaultFormatter = (value: number): string => value.toLocaleString()
+
+function useBillingTooltip(): {
+    ensureBillingTooltip: () => [Root, HTMLElement]
+    hideBillingTooltip: () => void
+} {
+    const tooltipElRef = useRef<HTMLElement | null>(null)
+    const tooltipRootRef = useRef<Root | null>(null)
+
+    const ensureBillingTooltip = useCallback((): [Root, HTMLElement] => {
+        if (!tooltipElRef.current) {
+            tooltipElRef.current = document.createElement('div')
+            tooltipElRef.current.id = 'BillingTooltipWrapper'
+            tooltipElRef.current.className =
+                'BillingTooltipWrapper hidden absolute z-10 p-2 bg-bg-light rounded shadow-md text-xs pointer-events-none border border-border'
+            document.body.appendChild(tooltipElRef.current)
+        }
+        if (!tooltipRootRef.current) {
+            tooltipRootRef.current = createRoot(tooltipElRef.current)
+        }
+        return [tooltipRootRef.current, tooltipElRef.current]
+    }, [])
+
+    const hideBillingTooltip = useCallback((): void => {
+        if (tooltipElRef.current) {
+            tooltipElRef.current.classList.add('hidden')
+            tooltipElRef.current.classList.remove('block')
+        }
+    }, [])
+
+    useEffect(
+        () => () => {
+            if (tooltipRootRef.current) {
+                setTimeout(() => tooltipRootRef.current?.unmount(), 0)
+            }
+            tooltipElRef.current?.remove()
+        },
+        []
+    )
+
+    return { ensureBillingTooltip, hideBillingTooltip }
+}
+
+export function SeriesColorDot({ colorIndex }: { colorIndex: number }): JSX.Element {
+    return <div className={`series-color-dot series-color-dot-${colorIndex % 15}`} />
+}
+
+export function BillingLineGraph({
```

Comment:
Not super familiar with our charting so I'm not looking at these charts too deeply other than testing them.